pkg/kubelet/kubelet.go

```diff
@@ -796,8 +803,21 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
 
 	klet.admitHandlers.AddPodAdmitHandler(klet.containerManager.GetAllocateResourcesPodAdmitHandler())
 
-	criticalPodAdmissionHandler := preemption.NewCriticalPodAdmissionHandler(klet.GetActivePods, killPodNow(klet.podWorkers, kubeDeps.Recorder), kubeDeps.Recorder)
-	klet.admitHandlers.AddPodAdmitHandler(lifecycle.NewPredicateAdmitHandler(klet.getNodeAnyWay, criticalPodAdmissionHandler, klet.containerManager.UpdatePluginResources))
+	// Any admission handler that depends on GetNode() (or getNodeAway()) must have its addition delayed
+	// after the node sync with the apiserver. This is done to prevent caching of Node objects.
+	go func() {
+		klog.Info("Waiting for node sync before adding a predicate admission handler")
+		for {
+			if nodeHasSynced() {
+				break
+			}
+			time.Sleep(waitForAPIServerSyncPeriod)
+		}
+		klog.Info("Adding predicate admission handler")
+		criticalPodAdmissionHandler := preemption.NewCriticalPodAdmissionHandler(klet.GetActivePods, killPodNow(klet.podWorkers, kubeDeps.Recorder), kubeDeps.Recorder)
+		klet.admitHandlers.AddPodAdmitHandler(lifecycle.NewPredicateAdmitHandler(klet.getNodeAnyWay, criticalPodAdmissionHandler, klet.containerManager.UpdatePluginResources))
```

Comment:
Is there a risk of hitting a concurrency issue by sticking this in a goroutine? We add more pod admin handlers on L811/812 (now 831/832), is it possible that this will be waiting when we hit those calls? Should we lift those into here too?

https://github.com/kubernetes/kubernetes/blob/3c514ae588fdd2867eb9cced1c43c9ee2276a7ce/pkg/kubelet/kubelet.go#L808-L809

Reply:
IMO, the new go routine start can be moved after all `AddPodAdmitHandler` calls in the main thread.
i.e. after klet.softAdmitHandlers.AddPodAdmitHandler(lifecycle.NewProcMountAdmitHandler(klet.containerRuntime))